### PR TITLE
Lazily-initialize the controllers and views.

### PR DESCRIPTION
**Pull request by justindarc**

### CODE REVIEW

File: app/js/controllers/main.js [hunk 1]
```diff
@@ -1,10 +1,28 @@
...
+/*global Controller*/
+/*global Gesture*/
 
 export default class MainController extends Controller {
   constructor(options) {
     super(options);
 
     this._waitToBeOpened();
+
+    this._loadedModules = false;
```
Comment (DouglasSherk) [inline app/js/controllers/main.js]: This pattern is very not-Promisey. I find this much clearer, and less prone to error:

``` js
this._initialized = new Promise((resolve, reject) => {
  // eventually
  resolve();
});

// ...

this._initialized.then(() => {
  // do something that requires initialization
});

// ...

getSomeAsyncValueThatRequiresInit() {
  return new Promise((resolve, reject) => {
    this._initialized.then(() => {
      // do stuff
      resolve();
    });
  });
}
```

File: app/js/controllers/main.js [hunk 2]
```diff
@@ -35,11 +53,82 @@ export default class MainController extends Controller {
...
+        return;
+      }
+
+      var editView = new EditView();
+      var actionMenuView = new ActionMenuView();
+      var settingsView = new SettingsView();
+      var viewSourceView = new ViewSourceView();
+      var appendChildView = new AppendChildView();
+      var moveView = new MoveView();
+      var mainView = new MainView({
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:71]: I really don't like how gluey this has become. You're just moving code around, so I'm not going to make it an action item, but we should really think about this. We originally used this pattern of passing controllers and views into themselves and each other because we wanted to be able to make views interchangeable within controllers. But we've only needed that in one case with the Sharing app, and never that I know of within the Customizer. So I'd say that this pattern optimizes for a use case that is rare for us. Maybe we should more deeply link controllers and views, e.g. controllers initialize their own view.

File: app/js/controllers/main.js [hunk 3]
```diff
@@ -35,11 +53,82 @@ export default class MainController extends Controller {
...
+
+      this.actionMenuController = actionMenuController;
+      this.settingsController = settingsController;
+
+      editController.mainController = this;
+      viewSourceController.mainController = this;
+      actionMenuController.mainController = this;
+      settingsController.mainController = this;
+      appendChildController.mainController = this;
+      moveController.mainController = this;
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:119]: Why not just use `window.mainController` instead? Again, we're moving code around, so I'm not going to make this an action either, but it's a good potential refactor.

File: app/js/controllers/main.js [hunk 4]
```diff
@@ -35,11 +53,82 @@ export default class MainController extends Controller {
...
+      resolve();
+    });
+  }
+
   // These private _wait methods call each other. If you try to call them
   // directly, you'll end up with more than one gesture detector listening
   // at a time.
   _waitToBeOpened() {
     Gesture.detect(this.openGesture)
+      .then(() => this._lazyLoadModules())
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:133]: Very slick, I like this.

File: _custombuild [hunk 5]
```diff
@@ -8,7 +8,10 @@
  */
 var gulp = require('gulp');
 var merge = require('gulp-merge');
+var uglify = require('gulp-uglify');
```
Comment (DouglasSherk) [inline _custombuild]: Are you actually using this anywhere?

File: _custombuild [hunk 6]
```diff
@@ -8,7 +8,10 @@
...
 var replace = require('gulp-replace');
+var wrapper = require('gulp-wrapper');
+var stripComments = require('gulp-strip-comments');
```
Comment (DouglasSherk) [inline _custombuild]: Are you actually using this anywhere?

Comment (justindarc) [inline _custombuild]: Oops. I forgot to remove those. I was trying to find the best way to get rid of comments which ended up just being a regex.

File: app/js/controllers/main.js [hunk 7]
```diff
@@ -35,11 +51,86 @@ export default class MainController extends Controller {
     };
   }
 
+  _lazyLoadModules() {
+    this._loadedModules = this._loadedModules || new Promise((resolve, reject) => {
+      /*jshint evil:true*/
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:56]: What is this needed for?

File: app/js/controllers/main.js [hunk 8]
```diff
@@ -35,11 +51,86 @@ export default class MainController extends Controller {
...
+      var source = window.localStorage.getItem('__CUSTOMIZER__componentsSource');
```
Comment (DouglasSherk) [inline app/js/controllers/main.js:57]: This could really use a comment.

File: app/js/controllers/main.js (hunk 7 above)
Comment (justindarc) [inline app/js/controllers/main.js:56]: jshint complains about using `eval()`

File: _custombuild [hunk 9]
```diff
@@ -40,11 +40,26 @@ var components = gulp.src([
   APP_ROOT + 'components/gaia-switch/gaia-switch.js',
   APP_ROOT + 'components/addon-generator/dist/addon-generator.js'
   ])
-  .pipe(concat('components-concat.js'));
+  .pipe(concat('components-concat.js'))
+  .pipe(replace(/gaia\-/g, 'customizer-gaia-'))
+  .pipe(replace(/customizer\-gaia\-icons\/customizer\-gaia\-icons\.css/g, 'gaia-icons/gaia-icons.css'))
+  .pipe(replace(/(?:\/\*(?:[\s\S]*?)\*\/)|(?:([\s;])+\/\/(?:.*)$)/gm, '')) // Strip comments
```
Comment (DouglasSherk) [inline _custombuild]: These could use comments explaining why they're needed.

File: _custombuild [hunk 10]
```diff
@@ -40,11 +40,26 @@ var components = gulp.src([
...
+  .pipe(wrapper({
+    header: 'function componentsSource() {/*',
+    footer: '*/}\n' +
+      '(function() {\n' +
+        'var source = componentsSource.toString();\n' +
+        'source = source.substr(31, source.length - 34);\n' +
```
Comment (DouglasSherk) [inline _custombuild]: Why is this needed?

File: _custombuild [hunk 11]
```diff
@@ -40,11 +40,26 @@ var components = gulp.src([
...
   APP_ROOT + 'components/addon-generator/dist/addon-generator.js'
   ])
-  .pipe(concat('components-concat.js'));
+  .pipe(concat('components-concat.js'))
+  .pipe(replace(/gaia\-/g, 'customizer-gaia-'))
+  .pipe(replace(/customizer\-gaia\-icons\/customizer\-gaia\-icons\.css/g, 'gaia-icons/gaia-icons.css'))
+  .pipe(replace(/(?:\/\*(?:[\s\S]*?)\*\/)|(?:([\s;])+\/\/(?:.*)$)/gm, '')) // Strip comments
+  .pipe(wrapper({
+    header: 'function componentsSource() {/*',
+    footer: '*/}\n' +
```
Comment (DouglasSherk) [inline _custombuild:61]: I think you could just use the ``` tag here, like this:

``` js
footer: `*/}
(function() {
    var source = componentsSource.toString();
    source = source.substr(31, source.length - 34);
    window.localStorage.setItem('__CUSTOMIZER__componentsSource', source);
})();
console.log('Injected components-concat.js');`,
// ...
```

Comment (justindarc) [inline _custombuild:61]: I'll try it and see if they're supported in this env.
